File: apps/android/app/src/main/kotlin/com/loonext/android/features/settings/ApiKeysSection.kt
package com.loonext.android.features.settings

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Checkbox
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.unit.dp
import com.loonext.android.core.i18n.LocalAppLocale
import com.loonext.android.core.i18n.t
import com.loonext.android.core.model.ApiKey
import com.loonext.android.core.model.ApiKeyList
import com.loonext.android.core.model.CreateApiKeyBody
import com.loonext.android.ui.common.CenteredError
import com.loonext.android.ui.common.LoadState
import com.loonext.android.ui.common.relativeTime
import com.loonext.android.ui.common.userMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch

/**
 * #243 — API keys. Parity with the web's /settings/api-keys.
 *
 * # Why the defaults here invert the ones next door
 *
 * Connections opens its form with every event ticked, because subscribing to
 * nothing is a mistake and eight empty boxes is where somebody gives up.
 *
 * This form opens with only the READ scopes ticked, and that is the same
 * principle reaching the opposite answer: a Smart Default is only smart when
 * being wrong about it is cheap. Being wrong about which events you receive
 * costs a redundant webhook. Being wrong about what a key can do costs
 * whatever the key can do.
 *
 * It is still a default rather than an empty form — reading is what a first
 * integration does, and somebody who needs writing knows they do.
 *
 * # The rest
 *
 * **"Last used" is the headline, not the creation date.** The question this
 * screen exists to answer is "can I safely switch this off", and that is the
 * only fact that answers it. It is repeated inside the confirmation when the
 * answer is "yes, recently".
 *
 * **Revoked keys stay in the list.** "What did we turn off, and when" is an
 * incident question a hiding list cannot answer.
 *
 * **The token is a card that stays on screen, not a snackbar.** It exists once
 * in the product's whole life, and on a phone a snackbar is gone in four
 * seconds — usually while the person is switching to the app they meant to
 * paste it into.
 *
 * *Applying: Smart Defaults (inverted, and said why), Loss Aversion, Ethical
 * Friction, Zen of Clarity.*
 */
@Composable
fun ApiKeysSection(scope: SettingsScope) {
    val locale = LocalAppLocale.current
    var refreshKey by remember { mutableIntStateOf(0) }
    var state by remember { mutableStateOf<LoadState<ApiKeyList>>(LoadState.Loading) }
    var creating by remember { mutableStateOf(false) }
    var minted by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(scope.companyId, refreshKey) {
        try {
            state = LoadState.Ready(scope.repo.apiKeys(scope.companyId))
        } catch (cause: CancellationException) {
            throw cause
        } catch (cause: Exception) {
            if (state is LoadState.Ready) scope.showMessage(cause.userMessage(locale))
            else state = LoadState.Failed(cause.userMessage(locale))
        }
    }

    when (val current = state) {
        is LoadState.Loading -> SettingsSectionSkeleton(cards = 1)
        is LoadState.Failed -> CenteredError(
            current.message,
            onRetry = { refreshKey++ },
            modifier = Modifier.padding(vertical = 48.dp),
        )

        is LoadState.Ready -> {
            val rows = current.value.keys
            val cap = current.value.cap
            val atCap = cap > 0 && current.value.live >= cap

            Column(verticalArrangement = Arrangement.spacedBy(16.dp)) {
                Text(
                    t("apiKeys.intro"),
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )

                minted?.let { token ->
                    TokenCard(token = token, onDone = { minted = null })
                }

                if (rows.isEmpty() && !creating) {
                    SettingsCard(title = t("apiKeys.empty")) {
                        Text(
                            t("apiKeys.emptyBody"),
                            style = MaterialTheme.typography.bodyMedium,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                        TextButton(onClick = { creating = true }) {
                            Text(t("apiKeys.createAction"))
                        }
                    }
                } else {
                    rows.forEach { key ->
                        ApiKeyCard(
                            scope = scope,
                            apiKey = key,
                            onChanged = { refreshKey++ },
                        )
                    }

                    if (creating) {
                        CreateApiKeyCard(
                            scope = scope,
                            onCancel = { creating = false },
                            onCreated = { token ->
                                creating = false
                                minted = token
                                refreshKey++
                            },
                        )
                    } else {
                        OutlinedButton(onClick = { creating = true }, enabled = !atCap) {
                            Text(t("apiKeys.createAction"))
                        }
                        if (atCap) {
                            Text(
                                t("apiKeys.capReached").replace("{count}", cap.toString()),
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                        }
                    }
                }

                Text(
                    t("apiKeys.developerNote"),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }
    }
}

/** The seven scopes, in the order the API promises them. */
private val SCOPES = listOf(
    "conversations:read",
    "messages:read",
    "messages:send",
    "contacts:read",
    "contacts:write",
    "tasks:read",
    "tasks:write",
)

/** The safe half, and what a first integration actually needs. */
private val DEFAULT_SCOPES = SCOPES.filter { it.endsWith(":read") }.toSet()

/** The catalogue key for a scope's human sentence. Mirrors the shared rule. */
private fun scopeLabelKey(scope: String): String {
    val camel = Regex(""":(\w)""").replace(scope) { it.groupValues[1].uppercase() }
    return "apiKeys.scope.$camel"
}

@Composable
private fun TokenCard(token: String, onDone: () -> Unit) {
    val clipboard = LocalClipboardManager.current
    var copied by remember { mutableStateOf(false) }
    SettingsCard(title = t("apiKeys.tokenTitle")) {
        Text(
            t("apiKeys.tokenBody"),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            token,
            style = MaterialTheme.typography.bodySmall.copy(fontFamily = FontFamily.Monospace),
        )
        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            OutlinedButton(onClick = {
                clipboard.setText(AnnotatedString(token))
                copied = true
            }) {
                Text(if (copied) t("apiKeys.tokenCopied") else t("apiKeys.tokenCopy"))
            }
            TextButton(onClick = onDone) { Text(t("apiKeys.tokenDone")) }
        }
    }
}

@Composable
private fun CreateApiKeyCard(
    scope: SettingsScope,
    onCancel: () -> Unit,
    onCreated: (String) -> Unit,
) {
    val locale = LocalAppLocale.current
    val coroutineScope = rememberCoroutineScope()
    var name by remember { mutableStateOf("") }
    // Smart Defaults, inverted. See the file header.
    var scopes by remember { mutableStateOf(DEFAULT_SCOPES) }
    var saving by remember { mutableStateOf(false) }

    // #228: composed OUTSIDE the coroutine — `t` is @Composable and cannot run
    // in one, so a sentence decided in a callback has to be hoisted or it
    // renders its own key.
    val needOneScope = t("apiKeys.needOneScope")

    SettingsCard(title = t("apiKeys.createTitle")) {
        OutlinedTextField(
            value = name,
            onValueChange = { name = it },
            label = { Text(t("apiKeys.nameLabel")) },
            placeholder = { Text(t("apiKeys.namePlaceholder")) },
            singleLine = true,
            modifier = Modifier.fillMaxWidth(),
        )

        Text(t("apiKeys.scopesLabel"), style = MaterialTheme.typography.titleSmall)
        Text(
            t("apiKeys.scopesHint"),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        SCOPES.forEach { scopeName ->
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier.fillMaxWidth(),
            ) {
                Checkbox(
                    checked = scopes.contains(scopeName),
                    onCheckedChange = { on ->
                        scopes = if (on) scopes + scopeName else scopes - scopeName
                    },
                )
                Text(
                    t(scopeLabelKey(scopeName)),
                    style = MaterialTheme.typography.bodyMedium,
                )
            }
        }

        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
            OutlinedButton(
                enabled = !saving && name.isNotBlank(),
                onClick = {
                    if (scopes.isEmpty()) {
                        scope.showMessage(needOneScope)
                        return@OutlinedButton
                    }
                    saving = true
                    coroutineScope.launch {
                        try {
                            val result = scope.repo.createApiKey(
                                scope.companyId,
                                CreateApiKeyBody(
                                    name = name.trim(),
                                    // Ordered by the promise, not by tap order,
                                    // so two workspaces that picked the same
                                    // scopes store the same list.
                                    scopes = SCOPES.filter { scopes.contains(it) },
                                ),
                            )
                            onCreated(result.tokenOnce)
                        } catch (cause: CancellationException) {
                            throw cause
                        } catch (cause: Exception) {
                            scope.showMessage(cause.userMessage(locale))
                        } finally {
                            saving = false
                        }
                    }
                },
            ) {
                Text(if (saving) t("apiKeys.savingAction") else t("apiKeys.saveAction"))
            }
            TextButton(onClick = onCancel) { Text(t("apiKeys.cancelAction")) }
        }
    }
}

@Composable
private fun ApiKeyCard(
    scope: SettingsScope,
    apiKey: ApiKey,
    onChanged: () -> Unit,
) {
    val locale = LocalAppLocale.current
    val coroutineScope = rememberCoroutineScope()
    var confirming by remember { mutableStateOf(false) }

    val revoked = apiKey.revokedAt != null

    SettingsCard(title = apiKey.name) {
        Text(
            "${apiKey.tokenPrefix}…",
            style = MaterialTheme.typography.bodySmall.copy(fontFamily = FontFamily.Monospace),
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        Text(
            when {
                revoked ->
                    t("apiKeys.revokedOn")
                        .replace("{when}", relativeTime(apiKey.revokedAt ?: ""))
                apiKey.lastUsedAt != null ->
                    t("apiKeys.lastUsed")
                        .replace("{when}", relativeTime(apiKey.lastUsedAt))
                else -> t("apiKeys.neverUsed")
            },
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
        apiKey.scopes.forEach { scopeName ->
            Text(
                t(scopeLabelKey(scopeName)),
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }

        if (!revoked) {
            TextButton(onClick = { confirming = true }) {
                Text(t("apiKeys.revokeAction"))
            }
        }
    }

    // Ethical Friction: this breaks a live integration and cannot be undone.
    if (confirming) {
        val body = buildString {
            append(t("apiKeys.revokeBody"))
            apiKey.lastUsedAt?.let { at ->
                append(" ")
                append(
                    t("apiKeys.revokeUsedWarning").replace("{when}", relativeTime(at)),
                )
            }
        }
        AlertDialog(
            onDismissRequest = { confirming = false },
            title = { Text(t("apiKeys.revokeTitle")) },
            text = { Text(body) },
            confirmButton = {
                TextButton(onClick = {
                    confirming = false
                    coroutineScope.launch {
                        try {
                            scope.repo.revokeApiKey(scope.companyId, apiKey.id)
                            onChanged()
                        } catch (cause: CancellationException) {
                            throw cause
                        } catch (cause: Exception) {
                            scope.showMessage(cause.userMessage(locale))
                        }
                    }
                }) { Text(t("apiKeys.revokeConfirm")) }
            },
            dismissButton = {
                TextButton(onClick = { confirming = false }) {
                    Text(t("apiKeys.keepIt"))
                }
            },
        )
    }
}

File: apps/ios/Loonext/Features/Settings/ApiKeysSection.swift
import SwiftUI

/// #243 — API keys. Parity with the web's /settings/api-keys and Android's
/// `ApiKeysSection.kt`.
///
/// # Why the defaults here invert the ones next door
///
/// Connections opens its form with every event ticked, because subscribing to
/// nothing is a mistake and eight empty boxes is where somebody gives up.
///
/// This form opens with only the READ scopes ticked, and that is the same
/// principle reaching the opposite answer: a Smart Default is only smart when
/// being wrong about it is cheap. Being wrong about which events you receive
/// costs a redundant webhook. Being wrong about what a key can do costs
/// whatever the key can do. It is still a default rather than an empty form —
/// reading is what a first integration does.
///
/// # The rest
///
/// - **"Last used" is the headline, not the creation date.** The question this
///   screen exists to answer is "can I safely switch this off", and that is
///   the only fact that answers it. It is repeated inside the confirmation
///   when the answer is "yes, recently". *Applying: Loss Aversion.*
/// - **Revoked keys stay in the list.** "What did we turn off, and when" is an
///   incident question a hiding list cannot answer.
/// - **The token is a card that stays on screen, not a toast.** It exists once
///   in the product's whole life, and on a phone a toast is gone in four
///   seconds — usually while the person is switching to the app they meant to
///   paste it into.
/// - **Revoking confirms.** It breaks a live integration and cannot be undone.
///   *Applying: Ethical Friction.*
@MainActor
struct ApiKeysSectionView: View {
    let scope: SettingsScope

    @State private var loaded = false
    @State private var failed: String?
    @State private var apiKeys: [ApiKey] = []
    @State private var cap = 0
    @State private var live = 0
    @State private var creating = false
    @State private var minted: String?

    @Environment(\.appLocale) private var appLocale

    private var atCap: Bool { cap > 0 && live >= cap }

    private func words(_ key: String) -> String {
        AppStrings.translate(appLocale, key)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text(words("apiKeys.intro"))
                .font(.footnote)
                .foregroundStyle(.secondary)

            if let minted {
                ApiTokenCard(token: minted) { self.minted = nil }
            }

            if let failed {
                SettingsCard(title: words("apiKeys.loadFailed")) {
                    Text(failed).font(.footnote).foregroundStyle(.secondary)
                    Button(words("apiKeys.createAction")) { Task { await load() } }
                        .font(.footnote)
                }
            } else if !loaded {
                Text(words("apiKeys.savingAction"))
                    .font(.footnote)
                    .foregroundStyle(.secondary)
            } else if apiKeys.isEmpty && !creating {
                SettingsCard(title: words("apiKeys.empty")) {
                    Text(words("apiKeys.emptyBody"))
                        .font(.footnote)
                        .foregroundStyle(.secondary)
                    Button(words("apiKeys.createAction")) { creating = true }
                        .font(.footnote)
                }
            } else {
                ForEach(apiKeys) { key in
                    ApiKeyCard(scope: scope, apiKey: key, onChanged: { Task { await load() } })
                }

                if creating {
                    CreateApiKeyCard(
                        scope: scope,
                        onCancel: { creating = false },
                        onCreated: { token in
                            creating = false
                            minted = token
                            Task { await load() }
                        }
                    )
                } else {
                    VStack(alignment: .leading, spacing: 4) {
                        Button(words("apiKeys.createAction")) { creating = true }
                            .font(.footnote)
                            .disabled(atCap)
                        if atCap {
                            Text(
                                AppStrings.translate(
                                    appLocale,
                                    "apiKeys.capReached",
                                    ["count": "\(cap)"]
                                )
                            )
                            .font(.caption)
                            .foregroundStyle(.secondary)
                        }
                    }
                }
            }

            Text(words("apiKeys.developerNote"))
                .font(.caption)
                .foregroundStyle(.secondary)
        }
        .task { await load() }
    }

    private func load() async {
        do {
            let list = try await scope.repo.apiKeys(scope.companyId)
            apiKeys = list.keys
            cap = list.cap
            live = list.live
            failed = nil
        } catch {
            failed = error.userMessage
        }
        loaded = true
    }
}

/// The seven scopes, in the order the API promises them.
let apiKeyScopeOrder = [
    "conversations:read",
    "messages:read",
    "messages:send",
    "contacts:read",
    "contacts:write",
    "tasks:read",
    "tasks:write",
]

/// The safe half, and what a first integration actually needs.
let apiKeyDefaultScopes = Set(apiKeyScopeOrder.filter { $0.hasSuffix(":read") })

/// The catalogue key for a scope's human sentence. Mirrors the shared rule.
func apiKeyScopeLabelKeyLocal(_ scope: String) -> String {
    var camel = ""
    var upperNext = false
    for character in scope {
        if character == ":" {
            upperNext = true
        } else if upperNext {
            camel.append(Character(character.uppercased()))
            upperNext = false
        } else {
            camel.append(character)
        }
    }
    return "apiKeys.scope.\(camel)"
}

@MainActor
private struct ApiTokenCard: View {
    let token: String
    let onDone: () -> Void

    @State private var copied = false
    @Environment(\.appLocale) private var appLocale

    var body: some View {
        SettingsCard(title: AppStrings.translate(appLocale, "apiKeys.tokenTitle")) {
            Text(AppStrings.translate(appLocale, "apiKeys.tokenBody"))
                .font(.footnote)
                .foregroundStyle(.secondary)
            Text(token)
                .font(.caption.monospaced())
                .textSelection(.enabled)
            HStack(spacing: 12) {
                Button(
                    AppStrings.translate(
                        appLocale,
                        copied ? "apiKeys.tokenCopied" : "apiKeys.tokenCopy"
                    )
                ) {
                    UIPasteboard.general.string = token
                    copied = true
                }
                .font(.footnote)
                Button(AppStrings.translate(appLocale, "apiKeys.tokenDone"), action: onDone)
                    .font(.footnote)
            }
        }
    }
}

@MainActor
private struct CreateApiKeyCard: View {
    let scope: SettingsScope
    let onCancel: () -> Void
    let onCreated: (String) -> Void

    @State private var name = ""
    /// Smart Defaults, inverted. See the file header.
    @State private var scopes = apiKeyDefaultScopes
    @State private var saving = false

    @Environment(\.appLocale) private var appLocale

    private func words(_ key: String) -> String {
        AppStrings.translate(appLocale, key)
    }

    var body: some View {
        SettingsCard(title: words("apiKeys.createTitle")) {
            VStack(alignment: .leading, spacing: 4) {
                Text(words("apiKeys.nameLabel")).font(.caption.weight(.medium))
                TextField(words("apiKeys.namePlaceholder"), text: $name)
                    .textFieldStyle(.roundedBorder)
            }

            VStack(alignment: .leading, spacing: 6) {
                Text(words("apiKeys.scopesLabel")).font(.caption.weight(.medium))
                Text(words("apiKeys.scopesHint"))
                    .font(.caption2)
                    .foregroundStyle(.secondary)
                ForEach(apiKeyScopeOrder, id: \.self) { scopeName in
                    Toggle(
                        words(apiKeyScopeLabelKeyLocal(scopeName)),
                        isOn: Binding(
                            get: { scopes.contains(scopeName) },
                            set: { on in
                                if on { scopes.insert(scopeName) } else { scopes.remove(scopeName) }
                            }
                        )
                    )
                    .font(.footnote)
                }
            }

            HStack(spacing: 12) {
                Button(words(saving ? "apiKeys.savingAction" : "apiKeys.saveAction")) {
                    Task { await save() }
                }
                .font(.footnote)
                .disabled(saving || name.trimmingCharacters(in: .whitespaces).isEmpty)
                Button(words("apiKeys.cancelAction"), action: onCancel)
                    .font(.footnote)
            }
        }
    }

    private func save() async {
        guard !scopes.isEmpty else {
            scope.showMessage(words("apiKeys.needOneScope"))
            return
        }
        saving = true
        defer { saving = false }
        do {
            let minted = try await scope.repo.createApiKey(
                scope.companyId,
                body: CreateApiKeyBody(
                    name: name.trimmingCharacters(in: .whitespaces),
                    // Ordered by the promise, not by tap order, so two
                    // workspaces that picked the same scopes store the same
                    // list.
                    scopes: apiKeyScopeOrder.filter { scopes.contains($0) }
                )
            )
            onCreated(minted.token_once)
        } catch {
            scope.showMessage(error.userMessage)
        }
    }
}

@MainActor
private struct ApiKeyCard: View {
    let scope: SettingsScope
    let apiKey: ApiKey
    let onChanged: () -> Void

    @State private var confirming = false
    @Environment(\.appLocale) private var appLocale

    private func words(_ key: String) -> String {
        AppStrings.translate(appLocale, key)
    }

    private var revoked: Bool { apiKey.revoked_at != nil }

    var body: some View {
        SettingsCard(title: apiKey.name) {
            Text("\(apiKey.token_prefix)…")
                .font(.caption.monospaced())
                .foregroundStyle(.secondary)

            Text(statusLine)
                .font(.caption)
                .foregroundStyle(.secondary)

            Divider()
            ForEach(apiKey.scopes, id: \.self) { scopeName in
                Text(words(apiKeyScopeLabelKeyLocal(scopeName)))
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            if !revoked {
                Button(words("apiKeys.revokeAction"), role: .destructive) {
                    confirming = true
                }
                .font(.footnote)
            }
        }
        // Ethical Friction: this breaks a live integration and cannot be undone.
        .alert(words("apiKeys.revokeTitle"), isPresented: $confirming) {
            Button(words("apiKeys.keepIt"), role: .cancel) {}
            Button(words("apiKeys.revokeConfirm"), role: .destructive) {
                Task { await revoke() }
            }
        } message: {
            Text(confirmBody)
        }
    }

    private var statusLine: String {
        if let revokedAt = apiKey.revoked_at {
            return AppStrings.translate(
                appLocale,
                "apiKeys.revokedOn",
                ["when": relativeTime(revokedAt)]
            )
        }
        if let usedAt = apiKey.last_used_at {
            return AppStrings.translate(
                appLocale,
                "apiKeys.lastUsed",
                ["when": relativeTime(usedAt)]
            )
        }
        return words("apiKeys.neverUsed")
    }

    private var confirmBody: String {
        var out = words("apiKeys.revokeBody")
        if let usedAt = apiKey.last_used_at {
            out += " "
            out += AppStrings.translate(
                appLocale,
                "apiKeys.revokeUsedWarning",
                ["when": relativeTime(usedAt)]
            )
        }
        return out
    }

    private func revoke() async {
        do {
            try await scope.repo.revokeApiKey(scope.companyId, id: apiKey.id)
            onChanged()
        } catch {
            scope.showMessage(error.userMessage)
        }
    }
}
